### ISSUE TITLE
Names and UB of popcount, find_first_set, find_last_set

### ISSUE THREAD

**Issue by mattkretz** [mattkretz/std-simd-feedback#1]
These need to be consistent and unambiguous with the corresponding functions introduced for builtin types (`countl_zero`, `countl_one`, `countr_zero`, `countr_one`, `popcount`). Note that there are two variants of these functions that are relevant to `simd<T>`:

1. Counting the number of entries in a `simd_mask<T>`.
  This returns a scalar integer or a `std::bitset`.
2. Element-wise counting the number of 1 bits in a `simd<T>`.
  This returns a `simd<int>`.

The former is present in Par TS 2. The latter should be added, now that the [bit.count] functions are present in C++20.

The `count[lr]_(zero|one)` functions in C++20 are slightly different from `find_(first|last)_set`. The latter return in what place the first `true` value in the mask appears. However, if there is none, the behavior is undefined.
The corresponding `count[lr]_zero` functions count the number of consecutive `false` values starting from the left or right and thus do not invoke undefined behavior.

**Comment by crtrott** [comment]
I prefer the count[lr]_(zero|one), its more useable/explainable, and its aligned with the rest of the standard.

**Comment by iburyl** [comment]
x86 has simd VPLZCNTD/Q defined without UB. Everything non expressible via this instruction will likely end up emulated in some close to scalar form, where fixing UB will not be a dominating overhead.

Thus, it is better to leave only functions with well defined behavior synced with <bits> header: drop find_first_set/find_last_set from spec and leave count[l/r]_[zero/one] with defined behavior in all cases.

**Comment by mattkretz** [comment]
What a can of worms. Please review Section 5.2 of [D1928R3.pdf](https://github.com/mattkretz/std-simd-feedback/files/10545315/D1928R3.pdf).

TL;DR: I object to the `count[lr]` names. Use `reduce_min_index` and `reduce_max_index` instead.

@danieltowner

**Comment by danieltowner** [comment]
I agree that the countlr names aren't great but they are what we have got already and changing them just introduces more confusion. I'm not sure it is up to simd_mask to try to fix mistakes in naming made elsewhere. I think my view is biased by always thinking about simd masks as a set of contiguous bits which are conventionally read MSB to LSB like the bits in any other integer, so the terms left and right seem reasonable in that context.

If we did rename those functions, why do they need the reduce prefix on them: just ``count``,  ``min_index``,  and ``max_index`` might suffice.

**Comment by mattkretz** [comment]
So assuming we go with `count[lr]_zero` to replace `find_(first|last)_set`:
```c++
void f(auto x) {
  auto a = countr_zero(x);
  auto b = countr_zero(x > 0);
  auto c = countr_zero(+(x > 0));
}
f(1); // a is int, b is int, c is int
f(simd<int>(1)); // a is simd<int>, b is int, c is simd<int>
```
This is not consistent. In one case `countr_zero` applies element-wise, in the other case it's a reduction.

One of the guiding principles of the `std::simd` API is that instructions/implementation does not guide the interface. The API design follows the idea of expressing data-parallelism via the type system. So it's irrelevant how the CPU determines the minimum index of the first `true` mask element.

So which one is more self-documenting:
```c++
int find(char needle, const std::string& hay) {
  using V = std::simd<char>;
  for (std::size_t i = 0; i + V::size <= hay.size(); i += V::size) {
    const V chunk(hay.data() + i);
    if (any_of(chunk == needle)) {
      return i + countr_zero(chunk == needle);
    }
  }
}
```
or
```c++
int find(char needle, const std::string& hay) {
  using V = std::simd<char>;
  for (std::size_t i = 0; i + V::size <= hay.size(); i += V::size) {
    const V chunk(hay.data() + i);
    if (any_of(chunk == needle)) {
      return i + reduce_min_index(chunk == needle);
    }
  }
}
```
? I'm open to better name suggestions. But I'm strongly opposed to using `count[lr]*` for *reduction* operations. I'm not insisting on calling them `reduce_*`. But it would be explicit about this operation not applying element-wise. I.e. consistency: "every operation applies element-wise, unless it's called `reduce*`¹".

Alternative name ideas: `minimum_index(chunk == needle)`, `first_index(chunk == needle)`

¹ or all_of, any_of, none_of (because these are very common and are reductions in other contexts as well)

**Comment by mattkretz** [comment]
Renamed for R4.
